### PR TITLE
GMX Arbitrum Abstractions Updates

### PR DESCRIPTION
**Pull request by 1chioku**
Brief comments on the purpose of your changes:
1. Updated gmx_arbitrum_glp_fees.sql to weth_fee_generated in addition to cumulative fee generated.
2. Added gmx_arbitrum_glp_float.sql for abstracting GLP mints and burns.
3. Updated gmx_arbitrum_glp_schema.yml and gmx_arbitrum_glp_sources.yml to reflect changes in 1 and 2.

*For Dune Engine V2*
I've checked that:
General checks:
* [x] I tested the query on dune.com after compiling the model with dbt compile (compiled queries are written to the target directory)
* [x] I used "refs" to reference other models in this repo and "sources" to reference raw or decoded tables 
* [x] if adding a new model, I added a test
* [x] the filename is unique and ends with .sql
* [x] each sql file is a select statement and has only one view, table or function defined  
* [x] column names are `lowercase_snake_cased`
* [x] if adding a new model, I edited the dbt project YAML file with new directory path for both models and seeds (if applicable)
* [x] if wanting to expose a model in the UI (Dune data explorer), I added a post-hook in the JINJA config to add metadata (blockchains, sector/project, name and contributors)
* [x] if adding a new materialized table, I edited the optimize table macro

Join logic:
* [ ] if joining to base table (i.e. ethereum transactions or traces), I looked to make it an inner join if possible

Incremental logic:
* [ ] I used is_incremental & not is_incremental jinja block filters on both base tables and decoded tables
  * [ ] where block_time >= date_trunc("day", now() - interval '1 week')
* [ ] if joining to base table (i.e. ethereum transactions or traces), I applied join condition where block_time >= date_trunc("day", now() - interval '1 week')
* [ ] if joining to prices view, I applied join condition where minute >= date_trunc("day", now() - interval '1 week')

### CODE REVIEW

**Comment by dot2dotseurat** [comment]
Hi @1chioku, 

It looks like we missed some things on the initial review of this PR. The materialization has been defined as "incremental" but we haven't added anything to actually make it incremental. Here are the general docs from dbt which will help orient but essentially we need to add a "is_incremental" block to tell dbt what data to use for an incremental run. https://docs.getdbt.com/docs/building-a-dbt-project/building-models/configuring-incremental-models

There's a range of examples but you can see one [here](https://github.com/duneanalytics/spellbook/blob/6a2be5ec1371f4d5bc4ad58872c8a4e70e4bb509/models/element/ethereum/element_ethereum_events.sql#L34) 

But in your case, since you are using a window function, you can't actually create an incremental table. But that may not have been necessary to begin with. 

I can see from our job orchestration tool that the table is being created in about 30s:
`15:17:05  42 of 46 OK created incremental model gmx_arbitrum.glp_fees .................... [OK in 34.91s]`

Could we make this a view instead of an incremental table? Otherwise, we can make this a table that is rebuilt every 30 minutes but perhaps it's better to make it a view so it's always up to date?

File: models/gmx/arbitrum/glp/gmx_arbitrum_glp_float.sql [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+        )
+}}
+
+WITH minute AS  -- This CTE generates a series of minute values
+    (
+    SELECT explode(sequence(TIMESTAMP '2021-08-31 08:13', CURRENT_TIMESTAMP, INTERVAL 1 minute)) AS minute -- 2021-08-31 08:13 is the timestamp of the first vault transaction
+    ) ,
+
+/*
+GLP tokens are minted and burned by the GLP Manager contract by invoking addLiquidity() and removeLiquidity()
```
Comment (dot2dotseurat) [inline models/gmx/arbitrum/glp/gmx_arbitrum_glp_float.sql:20]: good context for source yaml file

**Comment by dot2dotseurat** [comment]
Logic changes look good otherwise. Thank you for the detailed comments in the code. It would be nice to move some of the explanations into the schema file and source files so that information is available in the spellbook documentation. I leave a few notes where I think an inline comment would be better in the yaml files.

File: models/gmx/arbitrum/glp/gmx_arbitrum_glp_fees.sql [hunk 2]
```diff
@@ -21,53 +21,58 @@ GLP fees accrued to the Fee GLP contract and can be claimed by invoking function
...
-                AND `contract_address` = '0x82af49447d8a07e3bd95bd0d56f35241523fbab1' -- WETH Arbitrum Smart Contract
-                ) a
-            GROUP BY a.minute
-            ) b
-        ) c
+            SELECT -- This subquery truncates the block time to a minute and selects all the inbound tranfers of WETH tokens to the Fee GLP contract
+                date_trunc('minute', evt_block_time) AS minute,
+                ((value) / 1e18) AS transfer_value -- WETH 18dp
+            FROM {{ source('erc20_arbitrum', 'evt_transfer') }}
+            WHERE `to` = '0x4e971a87900b931ff39d1aad67697f49835400b6' -- Fee GLP contract
```
Comment (dot2dotseurat) [inline models/gmx/arbitrum/glp/gmx_arbitrum_glp_fees.sql:41]: these could be set as variables and you could share them across all the glp spells https://docs.getdbt.com/docs/building-a-dbt-project/building-models/using-variables

File: models/gmx/arbitrum/glp/gmx_arbitrum_glp_fees.sql [hunk 3]
```diff
@@ -21,53 +21,58 @@ GLP fees accrued to the Fee GLP contract and can be claimed by invoking function
...
+        ON a.minute = b.minute
+    LEFT JOIN
+        (
+        SELECT
+            minute,
+            weth_current_price
+        FROM {{ref('gmx_arbitrum_glp_components')}}
+        ) c 
+        ON a.minute = c.minute
+    ) x
```
Comment (dot2dotseurat) [inline models/gmx/arbitrum/glp/gmx_arbitrum_glp_fees.sql:78]: not gonna lie, I would love to see more meaningful subquery names. I think single letter subqueries are hard for other developers to interpret. I'll leave that for you to decide if you want to clean up.

**Comment by 1chioku** [comment]
@dot2dotseurat I put all my comments in this single reply.

**1. On the use of incremental model:**

- Agreed, for now some of the incremental tables can actually be converted into views, gmx_arbitrum_glp_fees.sql and gmx_arbitrum_glp_float.sql as such, this will done in the upcoming patch
- The reason why I was trying to push for incremental over views is that the data is updated every minute and over time the data table gets heavier; and I have plans for the tables will be used in the near future for higher level abstractions, so the 30 second run time for the view table will compound in higher runtime for the child tables later on.

**2. On the use of variables:**

- Thanks for the tip, I will apply same in the upcoming patch.

**3. On subquery naming:**

- These are actually made for my own personal usage, I use abc & xyz to track nesting orders, its particular useful is situations like these:
https://github.com/duneanalytics/spellbook/blob/main/models/gmx/arbitrum/glp/gmx_arbitrum_glp_components.sql#L836

- Unless necessary I will retain the same 

**4. On yaml files literature:**

- My plan is to make a mega schema update for the yaml files once I finish pushing all the immediate abstractions for the project, the comments are reminders for myself when that time comes.
- If you have any best practice example for rich schema writing, please send me link to same, I've been sneak peaking at what my fellow wizards are writing in their yaml files and it's pretty frugal.

**Comment by 1chioku** [comment]
@dot2dotseurat 

Thanks for the helpful comments, for now I'd like to have fee and float merged as views so I can go on with the testing.

I still have much to push for that specific project, and will implement the other recommendations then, baby steps for now.

**Review by dot2dotseurat — APPROVED** [review]

**Pull request merged**